packages/debugger/lib/data/reducers.js

```diff
@@ -165,31 +247,59 @@ function assignments(state = DEFAULT_ASSIGNMENTS, action) {
       debug("action.type %O", action.type);
       debug("action.assignments %O", action.assignments);
       return Object.values(action.assignments).reduce((acc, assignment) => {
-        let { id, astId, compilationId } = assignment;
+        let { id, astId, sourceAndPointer, compilationId } = assignment;
         //we assume for now that only ordinary variables will be assigned this
         //way, and not globals; globals are handled in DEFAULT_ASSIGNMENTS
-        return {
-          ...acc,
-          byId: {
-            ...acc.byId,
-            [id]: assignment
-          },
-          byCompilationId: {
-            ...acc.byCompilationId,
-            [compilationId]: {
-              byAstId: {
-                ...(acc.byCompilationId[compilationId] || {}).byAstId,
-                [astId]: [
-                  ...new Set([
-                    ...((acc.byCompilationId[compilationId] || { byAstId: {} })
-                      .byAstId[astId] || []),
-                    id
-                  ])
-                ]
+        if (astId !== undefined) {
+          return {
+            ...acc,
+            byId: {
+              ...acc.byId,
+              [id]: assignment
+            },
+            byCompilationId: {
+              ...acc.byCompilationId,
+              [compilationId]: {
+                ...acc.byCompilationId[compilationId],
+                byAstId: {
+                  ...(acc.byCompilationId[compilationId] || {}).byAstId,
+                  [astId]: [
+                    ...new Set([
+                      ...(((acc.byCompilationId[compilationId] || {}).byAstId ||
+                        {})[astId] || []),
+                      id
+                    ])
+                  ]
+                }
               }
             }
-          }
-        };
+          };
+        } else {
+          return {
+            ...acc,
+            byId: {
+              ...acc.byId,
+              [id]: assignment
+            },
+            byCompilationId: {
+              ...acc.byCompilationId,
+              [compilationId]: {
+                ...acc.byCompilationId[compilationId],
+                bySourceAndPointer: {
```

Comment:
As discussed, let's unify the state here to treat `sourceAndPointer` just like a regular opaque ID.

From that, we can unify the declare actions to take whichever info upfront (id vs. source/pointer)